### PR TITLE
Adding C++20 barrier

### PR DESCRIPTION
**Pull request by hkaiser**
- adding hpx/barrier.hpp that imports lcos::local::cpp20_barrier into namespace hpx

### CODE REVIEW

File: libs/synchronization/include/hpx/synchronization/barrier.hpp [hunk 1]
```diff
@@ -15,11 +15,213 @@
...
+    ///
+    /// The default value of the CompletionFunction template parameter is an
+    /// unspecified type, such that, in addition to satisfying the requirements
+    /// of CompletionFunction, it meets the Cpp17DefaultConstructible
+    /// requirements (Table 27) and completion() has no effects.
+    ///
+    /// barrier::arrival_token is an unspecified type, such that it meets the
+    /// Cpp17MoveConstructible (Table 28), Cpp17MoveAssignable (Table 30), and
+    /// Cpp17Destructible (Table 32) requirements.
+
```
Comment (msimberg) [inline libs/synchronization/include/hpx/synchronization/barrier.hpp:88]: Do you know if doxygen will take the docs above into account with the empty line (same with the other empty lines further up)?

**Comment by hkaiser** [comment]
> You've been busy, thanks! Do you know if we have any internal uses of barrier that should be replaced by the new one?

I wanted to go through our code in a second sweep to replace thing where possible.

File: libs/synchronization/include/hpx/synchronization/barrier.hpp (hunk 1 above)
Comment (hkaiser) [reply libs/synchronization/include/hpx/synchronization/barrier.hpp:88]: I shouldn't have marked all the standardeese as doxygen comments in the first place.

**Comment by hkaiser** [comment]
This should be good to go now.

**Pull request merged**